js/editor-controller.js

```diff
@@ -0,0 +1,235 @@
+     * 
+     * @param {object} editorActions - Functions needed to control the editor:
+     *      - setCode() - Replace the code in the editor.
+     *      - getCode() - Retrieve the code from the editor.
+     *      - onCodeChange() - Callback to run on every call change.
+     *      - loadHex() - Sending a hex into the editor to load it.
+     *      - loadFileToFs() - Sending a file to the editor to add it to the
+     *                         filesystem.
+     *      - setMobileEditor() - Function to run to set up the editor in a
+     *                            a special mode for the mobile apps.
+     */
+    _returnObj.setup = function(editorActions) {
+        _editorActions = editorActions;
+
+        window.addEventListener('message', function(event) {
+            if (!event.data) return;
+
+            if (event.data.type === CONTROLLER_MESSAGING.type) {
+                switch (event.data.action) {
+                    // Parent is sending code to update editor
+                    case CONTROLLER_MESSAGING.actions.importproject:
+                        if (!event.data.project || typeof event.data.project !== 'string') {
+                            throw new Error("Invalid 'project' data type. String should be provided.");
+                        }
+                        _editorActions.setCode(event.data.project);
+                        break;
+
+                    // Parent is sending initial code for editor
+                    // Also here we can sync parent data with editor's data
+                    case CONTROLLER_MESSAGING.actions.workspacesync:
+                        if (!event.data.projects || !Array.isArray(event.data.projects)) {
+                            throw new Error("Invalid 'projects' data type. Array should be provided.");
+                        }
+                        if (event.data.projects.length < 1) {
+                            throw new Error("'projects' array should contain at least one item.");
+                        }
+                        _editorActions.setCode(event.data.projects[0]);
+                        // Notify parent about editor successfully configured
+                        window.parent.postMessage({
+                            type: CONTROLLER_MESSAGING.type,
+                            action: CONTROLLER_MESSAGING.actions.workspaceloaded
+                        }, '*');
+                        break;
+
+                    // Parent is sending HEX file
+                    case CONTROLLER_MESSAGING.actions.loadhex:
+                        _editorActions.loadHex(event.data.filename, event.data.hexstring);
+                        break;
+
+                    // Parent is sending file for filesystem
+                    case CONTROLLER_MESSAGING.actions.loadfile:
+                        _editorActions.loadFileToFs(event.data.filename, event.data.filestring);
+                        break;
```

Comment:
In what cases will this be sent by the apps?
If this save a file into the filesystem, but I assume the apps probably want to send a main.py file (in whatever filename) and load the code into the editor instead of just the fs?

Reply:
The main feature I had in mind was loading a hex into the editor.

Reply:
I'd originally suggested it as `WebView` in Android doesn't support `<input type="file"`, and would need an alternate way to upload files - switching to `WebChromeView` will support file uploads

I think it's still a useful feature to include, as will give us the possibility of using an in-app native text editor in the future

Reply:
> The main feature I had in mind was loading a hex into the editor.

Right, for that we have the `loadhex` message. This would be more about sending some Python code into the editor.



> I'd originally suggested it as `WebView` in Android doesn't support `<input type="file"`, and would need an alternate way to upload files - switching to `WebChromeView` will support file uploads

Oh yes, I remember that now. I guess the other thing related to this would be if the WebView/WebChromeView has access to the same files that the app and/or chrome do?
If within the app a user downloads a Python file, where can that file end up? And can a user load a file downloaded from either the app or the browser?
This question is also relevant for the iOS app, as iOS has more restrictions around this.

Reply:
Our plan is that, as for MakeCode, hex files will be stored in the app, while other files will trigger a Share dialog. Hex files stored in the app can be copied out via Share. MakeCode/Home/Import can see files from any location.

Reply:
Okay so we have:
-  `loadhex` to load a hex file into the editor
- `loadfile` to add a file to the filesystem (from the previous PRs)
    - This will ad a file to the filesystem, not the code editor
    - Unless it's a file named `main.py` in which case it will ask before loading it to the code editor

I would suggest to change `loadfile` to from using `loadFileToFilesystem()` to use `loadPy()` instead, which does the following:
- Load the python code into the editor
- Unless it has the magic comment that makes it load only into the filesystem
- This is the same as in a drag&drop a python file into the editor (from a PC desktop environment)

Sounds like `CONTROLLER_MESSAGING.actions.loadfile` might not be used for the time being, but this feels like a good default?

Reply:
👍